File: python_firmware/w5500_driver.py
import machine
import time

class W5500:
    """
    W5500-EVB-Pico2 전용 순수 MicroPython 드라이버 (TCP + UDP 멀티 소켓 지원)
    Socket 0: TCP Web Server (Port 80)
    Socket 1: UDP SNMP Agent (Port 161)
    """
    # Block Select Control Bytes for S0 and S1
    COMMON_REG_R = 0x00
    COMMON_REG_W = 0x04
    
    # Socket 0 Control Bytes (BSB 1)
    S0_REG_R     = 0x08
    S0_REG_W     = 0x0C
    S0_TX_W      = 0x14
    S0_RX_R      = 0x18

    # Socket 1 Control Bytes (BSB 2)
    S1_REG_R     = 0x28
    S1_REG_W     = 0x2C
    S1_TX_W      = 0x34
    S1_RX_R      = 0x38

    # Commands (Sn_CR)
    CR_OPEN      = 0x01
    CR_LISTEN    = 0x02
    CR_CONNECT   = 0x04
    CR_DISCON    = 0x08
    CR_CLOSE     = 0x10
    CR_SEND      = 0x20
    CR_RECV      = 0x40

    # Status (Sn_SR)
    SOCK_CLOSED      = 0x00
    SOCK_INIT        = 0x13
    SOCK_LISTEN      = 0x14
    SOCK_ESTABLISHED = 0x17
    SOCK_CLOSE_WAIT  = 0x1C
    SOCK_UDP         = 0x22

    def __init__(self, spi, cs_pin, rst_pin=20):
        self.spi = spi
        self.cs = machine.Pin(cs_pin, machine.Pin.OUT)
        self.cs.value(1)
        
        if rst_pin is not None:
            rst = machine.Pin(rst_pin, machine.Pin.OUT)
            rst.value(0)
            time.sleep_ms(50)
            rst.value(1)
            time.sleep_ms(100)

    def _write_common(self, addr, data):
        self.cs.value(0)
        buf = bytearray([(addr >> 8) & 0xFF, addr & 0xFF, self.COMMON_REG_W])
        buf.extend(data)
        self.spi.write(buf)
        self.cs.value(1)

    def _read_common(self, addr, length):
        self.cs.value(0)
        buf = bytearray([(addr >> 8) & 0xFF, addr & 0xFF, self.COMMON_REG_R])
        self.spi.write(buf)
        res = self.spi.read(length)
        self.cs.value(1)
        return res

    def _write_sn(self, sn, addr, data):
        reg_w = self.S0_REG_W if sn == 0 else self.S1_REG_W
        self.cs.value(0)
        buf = bytearray([(addr >> 8) & 0xFF, addr & 0xFF, reg_w])
        buf.extend(data)
        self.spi.write(buf)
        self.cs.value(1)

    def _read_sn(self, sn, addr, length):
        reg_r = self.S0_REG_R if sn == 0 else self.S1_REG_R
        self.cs.value(0)
        buf = bytearray([(addr >> 8) & 0xFF, addr & 0xFF, reg_r])
        self.spi.write(buf)
        res = self.spi.read(length)
        self.cs.value(1)
        return res

    def _cmd_sn(self, sn, cmd):
        self._write_sn(sn, 0x0001, bytes([cmd]))
        for _ in range(100):
            res = self._read_sn(sn, 0x0001, 1)
            if res and len(res) > 0 and res[0] == 0:
                return
            time.sleep_us(10)

    def setup_network(self, ip, subnet="255.255.255.0", gateway="192.168.10.1", mac="00:08:DC:23:50:0A"):
        """네트워크 고정 IP 및 MAC 주소 설정"""
        try:
            mac_bytes = bytes([int(x, 16) for x in mac.split(':')])
            ip_bytes = bytes([int(x) for x in ip.split('.')])
            sn_bytes = bytes([int(x) for x in subnet.split('.')])
            gw_bytes = bytes([int(x) for x in gateway.split('.')])

            self._write_common(0x0001, gw_bytes)   # Gateway
            self._write_common(0x0005, sn_bytes)   # Subnet
            self._write_common(0x0009, mac_bytes)  # MAC
            self._write_common(0x000F, ip_bytes)   # Source IP

            # RTR (Retransmission Retry Time): 10,000 * 100us = 1.0초 (0x2710) - Flash 기록 중 TCP 윈도우 대기 보장
            self._write_common(0x0019, bytes([0x27, 0x10]))
            # RTY (Retry Count): 15회 (0x0F)
            self._write_common(0x001B, bytes([0x0F]))
            
            time.sleep_ms(10)
        except Exception as e:
            print("⚠️ W5500 네트워크 설정 예외:", e)

    def get_ip(self):
        try:
            res = self._read_common(0x000F, 4)
            if res and len(res) == 4:
                return ".".join(str(b) for b in res)
        except Exception:
            pass
        return "0.0.0.0"

    def is_link_up(self):
        """W5500 PHY 레지스터(0x002E) Bit 0 (LNK) 확인하여 케이블 물리적 연결 여부 반환"""
        try:
            res = self._read_common(0x002E, 1)
            if res and len(res) > 0:
                return bool(res[0] & 0x01)
        except Exception:
            pass
        return False


    # ----------------------------------------------------
    # Socket 0: TCP Server Methods (Web Server)
    # ----------------------------------------------------
    def close_socket(self, sn=0):
        """소켓 강제 닫기 (CR_CLOSE)"""
        try:
            self._cmd_sn(sn, self.CR_CLOSE)
            time.sleep_ms(5)
        except Exception:
            pass

    def listen_server(self, port=80):
        try:
            res = self._read_sn(0, 0x0003, 1)
            status = res[0] if res and len(res) > 0 else self.SOCK_CLOSED
            # 이미 LISTEN 또는 ESTABLISHED 인 경우는 유지하고, 그 외에는 닫고 재오픈
            if status != self.SOCK_LISTEN and status != self.SOCK_ESTABLISHED:
                self.close_socket(0)
                self._write_sn(0, 0x001E, bytes([0x0C])) # Socket 0 RX Buffer: 12KB (대용량 TCP 수신 버퍼)
                self._write_sn(0, 0x001F, bytes([0x02])) # Socket 0 TX Buffer: 2KB
                self._write_sn(0, 0x0000, bytes([0x01])) # TCP mode
                self._write_sn(0, 0x0004, bytes([(port >> 8) & 0xFF, port & 0xFF]))
                self._cmd_sn(0, self.CR_OPEN)
                time.sleep_ms(5)
                self._cmd_sn(0, self.CR_LISTEN)
                time.sleep_ms(5)
        except Exception as e:
            print("⚠️ listen_server 오류:", e)


    def get_socket_status(self, sn=0):
        try:
            res = self._read_sn(sn, 0x0003, 1)
            if res and len(res) > 0:
                return res[0]
        except Exception:
            pass
        return self.SOCK_CLOSED

    def rx_bytes_available(self, sn=0):
        try:
            res = self._read_sn(sn, 0x0026, 2)
            if res and len(res) >= 2:
                return (res[0] << 8) | res[1]
        except Exception:
            pass
        return 0

    def read_rx_data(self, sn=0):
        rx_size = self.rx_bytes_available(sn)
        if rx_size == 0:
            return b""
        
        try:
            rd_ptr_buf = self._read_sn(sn, 0x0028, 2)
            if not rd_ptr_buf or len(rd_ptr_buf) < 2:
                return b""
            rd_ptr = (rd_ptr_buf[0] << 8) | rd_ptr_buf[1]
            
            rx_r_cb = self.S0_RX_R if sn == 0 else self.S1_RX_R
            self.cs.value(0)
            buf = bytearray([(rd_ptr >> 8) & 0xFF, rd_ptr & 0xFF, rx_r_cb])
            self.spi.write(buf)
            data = self.spi.read(rx_size)
            self.cs.value(1)
            
            new_rd = (rd_ptr + rx_size) & 0xFFFF
            self._write_sn(sn, 0x0028, bytes([(new_rd >> 8) & 0xFF, new_rd & 0xFF]))
            self._cmd_sn(sn, self.CR_RECV)
            return data
        except Exception as e:
            self.cs.value(1)
            print("⚠️ read_rx_data 오류:", e)
            return b""

    def get_tx_free_size(self, sn=0):
        try:
            res = self._read_sn(sn, 0x0020, 2)
            if res and len(res) >= 2:
                return (res[0] << 8) | res[1]
        except Exception:
            pass
        return 0

    def send_tx_data(self, data, sn=0):
        data_len = len(data)
        if data_len == 0:
            return
            
        try:
            offset = 0
            chunk_size = 1024
            
            while offset < data_len:
                chunk = data[offset : offset + chunk_size]
                chunk_len = len(chunk)
                
                # TX Free Size 가 chunk_len 이상이 될 때까지 대기 (최대 100ms)
                for _ in range(50):
                    if self.get_tx_free_size(sn) >= chunk_len:
                        break
                    time.sleep_ms(2)
                
                wr_ptr_buf = self._read_sn(sn, 0x0024, 2)
                if not wr_ptr_buf or len(wr_ptr_buf) < 2:
                    return
                wr_ptr = (wr_ptr_buf[0] << 8) | wr_ptr_buf[1]
                
                tx_w_cb = self.S0_TX_W if sn == 0 else self.S1_TX_W
                self.cs.value(0)
                buf = bytearray([(wr_ptr >> 8) & 0xFF, wr_ptr & 0xFF, tx_w_cb])
                buf.extend(chunk)
                self.spi.write(buf)
                self.cs.value(1)
                
                new_wr = (wr_ptr + chunk_len) & 0xFFFF
                self._write_sn(sn, 0x0024, bytes([(new_wr >> 8) & 0xFF, new_wr & 0xFF]))
                self._cmd_sn(sn, self.CR_SEND)
                time.sleep_ms(5)
                
                offset += chunk_len
        except Exception as e:
            self.cs.value(1)
            print("⚠️ send_tx_data 오류:", e)

    def disconnect_socket(self, sn=0):
        try:
            self._cmd_sn(sn, self.CR_DISCON)
            time.sleep_ms(5)
            self._cmd_sn(sn, self.CR_CLOSE)
        except Exception:
            pass

    # ----------------------------------------------------
    # Socket 1: UDP Listener Methods (SNMP Agent - Port 161)
    # ----------------------------------------------------
    def open_udp_socket(self, sn=1, port=161):
        try:
            res = self._read_sn(sn, 0x0003, 1)
            status = res[0] if res and len(res) > 0 else self.SOCK_CLOSED
            if status != self.SOCK_UDP:
                self._cmd_sn(sn, self.CR_CLOSE)
                time.sleep_ms(2)
                self._write_sn(sn, 0x0000, bytes([0x02])) # UDP mode (0x02)
                self._write_sn(sn, 0x0004, bytes([(port >> 8) & 0xFF, port & 0xFF]))
                self._cmd_sn(sn, self.CR_OPEN)
                time.sleep_ms(5)
        except Exception as e:
            print("⚠️ open_udp_socket 오류:", e)

    def recv_udp_packet(self, sn=1):
        """W5500 UDP 헤더 [Remote IP 4B, Remote Port 2B, Data Len 2B] + Data 읽기"""
        rx_size = self.rx_bytes_available(sn)
        if rx_size < 8:
            return None, None, b""
            
        raw_data = self.read_rx_data(sn)
        if len(raw_data) < 8:
            return None, None, b""
            
        remote_ip = ".".join(str(b) for b in raw_data[0:4])
        remote_port = (raw_data[4] << 8) | raw_data[5]
        data_len = (raw_data[6] << 8) | raw_data[7]
        payload = raw_data[8:8+data_len]
        
        return remote_ip, remote_port, payload

    def send_udp_packet(self, remote_ip, remote_port, payload, sn=1):
        """지정한 상대방 IP 및 포트로 UDP 패킷 전송"""
        try:
            ip_bytes = bytes([int(x) for x in remote_ip.split('.')])
            self._write_sn(sn, 0x000C, ip_bytes) # Destination IP
            self._write_sn(sn, 0x0010, bytes([(remote_port >> 8) & 0xFF, remote_port & 0xFF])) # Destination Port
            self.send_tx_data(payload, sn=sn)
        except Exception as e:
            print("⚠️ send_udp_packet 오류:", e)

File: c_firmware/w5500_driver.c
#include "config.h"

#if (TARGET_ETH_CHIP == CHIP_W5500)

#include "w5500_driver.h"
#include "hardware/spi.h"
#include "hardware/gpio.h"
#include "pico/stdlib.h"
#include <string.h>
#include <stdio.h>

#define COMMON_REG_R 0x00
#define COMMON_REG_W 0x04

#define CR_OPEN      0x01
#define CR_LISTEN    0x02
#define CR_DISCON    0x08
#define CR_CLOSE     0x10
#define CR_SEND      0x20
#define CR_RECV      0x40
#define CR_SEND_MAC  0x21

static inline uint8_t get_sn_reg_r(uint8_t sn) { return ((sn * 4 + 1) << 3) | 0x00; }
static inline uint8_t get_sn_reg_w(uint8_t sn) { return ((sn * 4 + 1) << 3) | 0x04; }
static inline uint8_t get_sn_tx_w(uint8_t sn)  { return ((sn * 4 + 2) << 3) | 0x04; }
static inline uint8_t get_sn_rx_r(uint8_t sn)  { return ((sn * 4 + 3) << 3) | 0x00; }

static inline void cs_select(void) {
    while (spi_is_readable(SPI_PORT)) {
        uint8_t dummy;
        spi_read_blocking(SPI_PORT, 0, &dummy, 1);
    }
    gpio_put(PIN_CS, 0);
}

static inline void cs_deselect(void) {
    gpio_put(PIN_CS, 1);
}

static void write_common(uint16_t addr, const uint8_t *data, uint16_t len) {
    cs_select();
    uint8_t hdr[3] = { (uint8_t)(addr >> 8), (uint8_t)(addr & 0xFF), COMMON_REG_W };
    spi_write_blocking(SPI_PORT, hdr, 3);
    spi_write_blocking(SPI_PORT, data, len);
    cs_deselect();
}

static void read_common(uint16_t addr, uint8_t *buf, uint16_t len) {
    cs_select();
    uint8_t hdr[3] = { (uint8_t)(addr >> 8), (uint8_t)(addr & 0xFF), COMMON_REG_R };
    spi_write_blocking(SPI_PORT, hdr, 3);
    spi_read_blocking(SPI_PORT, 0x00, buf, len);
    cs_deselect();
}

static void write_sn(uint8_t sn, uint16_t addr, const uint8_t *data, uint16_t len) {
    uint8_t reg_w = get_sn_reg_w(sn);
    cs_select();
    uint8_t hdr[3] = { (uint8_t)(addr >> 8), (uint8_t)(addr & 0xFF), reg_w };
    spi_write_blocking(SPI_PORT, hdr, 3);
    spi_write_blocking(SPI_PORT, data, len);
    cs_deselect();
}

static void read_sn(uint8_t sn, uint16_t addr, uint8_t *buf, uint16_t len) {
    uint8_t reg_r = get_sn_reg_r(sn);
    cs_select();
    uint8_t hdr[3] = { (uint8_t)(addr >> 8), (uint8_t)(addr & 0xFF), reg_r };
    spi_write_blocking(SPI_PORT, hdr, 3);
    spi_read_blocking(SPI_PORT, 0x00, buf, len);
    cs_deselect();
}

static void cmd_sn(uint8_t sn, uint8_t cmd) {
    write_sn(sn, 0x0001, &cmd, 1);
    uint64_t start_t = time_us_64();
    while (time_us_64() - start_t < 100000ULL) {
        uint8_t reg_val = 0;
        read_sn(sn, 0x0001, &reg_val, 1);
        if (reg_val == 0) break;
        sleep_us(50);
    }
}

static void configure_socket_memory(void) {
    // Socket 0 (HTTP): 8KB TX, 8KB RX
    uint8_t s0_tx = 8, s0_rx = 8;
    write_sn(0, 0x001E, &s0_tx, 1);
    write_sn(0, 0x001F, &s0_rx, 1);

    // Socket 1 (SNMP): 4KB TX, 4KB RX
    uint8_t s1_tx = 4, s1_rx = 4;
    write_sn(1, 0x001E, &s1_tx, 1);
    write_sn(1, 0x001F, &s1_rx, 1);

    // Sockets 2-3: 2KB TX, 2KB RX (Socket 2: 2KB, Socket 3: 2KB -> Total sum 8+4+2+2 = 16KB)
    for (uint8_t sn = 2; sn <= 3; sn++) {
        uint8_t tx = 2, rx = 2;
        write_sn(sn, 0x001E, &tx, 1);
        write_sn(sn, 0x001F, &rx, 1);
    }

    // Sockets 4-7: 0KB TX, 0KB RX
    for (uint8_t sn = 4; sn < 8; sn++) {
        uint8_t tx = 0, rx = 0;
        write_sn(sn, 0x001E, &tx, 1);
        write_sn(sn, 0x001F, &rx, 1);
    }
}

void w5500_init(void) {
    spi_init(SPI_PORT, 25000000); // 25MHz High-Speed SPI clock
    gpio_set_function(PIN_SCK, GPIO_FUNC_SPI);
    gpio_set_function(PIN_MOSI, GPIO_FUNC_SPI);
    gpio_set_function(PIN_MISO, GPIO_FUNC_SPI);

    gpio_init(PIN_CS);
    gpio_set_dir(PIN_CS, GPIO_OUT);
    cs_deselect();

    gpio_init(PIN_RST);
    gpio_set_dir(PIN_RST, GPIO_OUT);
    gpio_put(PIN_RST, 0);
    sleep_ms(50);
    gpio_put(PIN_RST, 1);
    sleep_ms(100);

    // Issue W5500 Mode Register Software Reset (0x80) to clean internal registers
    uint8_t mr_rst = 0x80;
    write_common(0x0000, &mr_rst, 1);
    sleep_ms(50);

    uint8_t ver = 0;
    read_common(0x0039, &ver, 1);
    printf("📡 W5500 Version Register (0x0039): 0x%02X\n", ver);
    if (ver == 0x04) {
        printf("✅ W5500 Hardware Chip Detected Successfully! (ID: 0x04)\n");
    } else {
        printf("⚠️ W5500 Chip Read Warning: Expected 0x04, Got 0x%02X\n", ver);
    }

    configure_socket_memory();
}

void w5500_setup_network(const uint8_t ip[4], const uint8_t sn[4], const uint8_t gw[4], const uint8_t mac[6]) {
    write_common(0x0001, gw, 4);
    write_common(0x0005, sn, 4);
    write_common(0x0009, mac, 6);
    write_common(0x000F, ip, 4);

    // RTR: 1.0s (0x2710)
    uint8_t rtr[2] = { 0x27, 0x10 };
    write_common(0x0019, rtr, 2);
    // RTY: 15
    uint8_t rty = 0x0F;
    write_common(0x001B, &rty, 1);
    sleep_ms(10);
}

bool w5500_is_link_up(void) {
    uint8_t phy = 0;
    read_common(0x002E, &phy, 1);
    return (phy & 0x01) != 0;
}

void w5500_close_socket(uint8_t sn) {
    cmd_sn(sn, CR_CLOSE);
    sleep_us(100);
}

void w5500_listen_server(uint16_t port) {
    uint8_t status = w5500_get_socket_status(0);
    if (status != SOCK_LISTEN && status != SOCK_ESTABLISHED) {
        w5500_close_socket(0);
        uint8_t mode = 0x01; // TCP
        write_sn(0, 0x0000, &mode, 1);
        uint8_t port_buf[2] = { (uint8_t)(port >> 8), (uint8_t)(port & 0xFF) };
        write_sn(0, 0x0004, port_buf, 2);

        cmd_sn(0, CR_OPEN);
        for (int i = 0; i < 100; i++) {
            if (w5500_get_socket_status(0) == SOCK_INIT) break;
            sleep_us(100);
        }

        cmd_sn(0, CR_LISTEN);
        for (int i = 0; i < 100; i++) {
            if (w5500_get_socket_status(0) == SOCK_LISTEN) break;
            sleep_us(100);
        }
    }
}

uint8_t w5500_get_socket_status(uint8_t sn) {
    uint8_t status = SOCK_CLOSED;
    read_sn(sn, 0x0003, &status, 1);
    return status;
}

uint16_t w5500_rx_bytes_available(uint8_t sn) {
    uint16_t val1 = 0, val2 = 0;
    int retry = 50;
    do {
        uint8_t buf1[2] = {0}, buf2[2] = {0};
        read_sn(sn, 0x0026, buf1, 2);
        read_sn(sn, 0x0026, buf2, 2);
        val1 = (uint16_t)((buf1[0] << 8) | buf1[1]);
        val2 = (uint16_t)((buf2[0] << 8) | buf2[1]);
        if (--retry <= 0) break;
    } while (val1 != val2);
    return val1;
}

uint16_t w5500_read_rx_data(uint8_t sn, uint8_t *buf, uint16_t max_len) {
    uint16_t avail = w5500_rx_bytes_available(sn);
    if (avail == 0) return 0;

    uint16_t read_len = (avail > max_len) ? max_len : avail;
    uint8_t ptr_buf[2] = {0};
    read_sn(sn, 0x0028, ptr_buf, 2);
    uint16_t rd_ptr = (uint16_t)((ptr_buf[0] << 8) | ptr_buf[1]);

    // Socket buffer size mask (Socket 0: 8KB = 0x1FFF, Socket 1: 4KB = 0x0FFF, etc.)
    uint16_t buf_size = (sn == 0) ? 8192 : ((sn == 1) ? 4096 : 2048);
    uint16_t buf_mask = buf_size - 1;
    uint16_t offset = rd_ptr & buf_mask;
    uint8_t rx_r_cb = get_sn_rx_r(sn);

    if (offset + read_len <= buf_size) {
        // Continuous block read
        cs_select();
        uint8_t hdr[3] = { (uint8_t)(offset >> 8), (uint8_t)(offset & 0xFF), rx_r_cb };
        spi_write_blocking(SPI_PORT, hdr, 3);
        spi_read_blocking(SPI_PORT, 0x00, buf, read_len);
        cs_deselect();
    } else {
        // Wrap-around split read across 8KB circular ring boundary
        uint16_t size1 = buf_size - offset;
        uint16_t size2 = read_len - size1;

        cs_select();
        uint8_t hdr1[3] = { (uint8_t)(offset >> 8), (uint8_t)(offset & 0xFF), rx_r_cb };
        spi_write_blocking(SPI_PORT, hdr1, 3);
        spi_read_blocking(SPI_PORT, 0x00, buf, size1);
        cs_deselect();

        cs_select();
        uint8_t hdr2[3] = { 0x00, 0x00, rx_r_cb };
        spi_write_blocking(SPI_PORT, hdr2, 3);
        spi_read_blocking(SPI_PORT, 0x00, buf + size1, size2);
        cs_deselect();
    }

    uint16_t new_rd = rd_ptr + read_len;
    uint8_t new_rd_buf[2] = { (uint8_t)(new_rd >> 8), (uint8_t)(new_rd & 0xFF) };
    write_sn(sn, 0x0028, new_rd_buf, 2);
    cmd_sn(sn, CR_RECV);

    return read_len;
}

static uint16_t get_tx_free_size(uint8_t sn) {
    uint16_t val1 = 0, val2 = 0;
    int retry = 50;
    do {
        uint8_t buf1[2] = {0}, buf2[2] = {0};
        read_sn(sn, 0x0020, buf1, 2);
        read_sn(sn, 0x0020, buf2, 2);
        val1 = (uint16_t)((buf1[0] << 8) | buf1[1]);
        val2 = (uint16_t)((buf2[0] << 8) | buf2[1]);
        if (--retry <= 0) break;
    } while (val1 != val2);
    return val1;
}

static bool is_tx_buffer_empty(uint8_t sn) {
    uint8_t rd_buf1[2] = {0}, rd_buf2[2] = {0};
    uint8_t wr_buf1[2] = {0}, wr_buf2[2] = {0};
    int retry = 50;

    do {
        read_sn(sn, 0x0022, rd_buf1, 2);
        read_sn(sn, 0x0022, rd_buf2, 2);
        if (--retry <= 0) break;
    } while (rd_buf1[0] != rd_buf2[0] || rd_buf1[1] != rd_buf2[1]);

    retry = 50;
    do {
        read_sn(sn, 0x0024, wr_buf1, 2);
        read_sn(sn, 0x0024, wr_buf2, 2);
        if (--retry <= 0) break;
    } while (wr_buf1[0] != wr_buf2[0] || wr_buf1[1] != wr_buf2[1]);

    uint16_t rd = (uint16_t)((rd_buf1[0] << 8) | rd_buf1[1]);
    uint16_t wr = (uint16_t)((wr_buf1[0] << 8) | wr_buf1[1]);

    return (rd == wr);
}

void w5500_send_tx_data(uint8_t sn, const uint8_t *data, uint16_t len) {
    if (len == 0 || data == NULL) return;

    uint8_t tx_w_cb = get_sn_tx_w(sn);
    uint16_t sent = 0;
    uint16_t buf_size = (sn == 0) ? 8192 : ((sn == 1) ? 4096 : 2048);
    uint16_t buf_mask = buf_size - 1;

    while (sent < len) {
        uint16_t free_size = 0;
        uint64_t start_t = time_us_64();
        while (1) {
            free_size = get_tx_free_size(sn);
            if (free_size > 0) break;
            if (time_us_64() - start_t > 3000000ULL) {
                printf("⚠️ [W5500] Socket %d TX buffer timeout\n", sn);
                return;
            }
            sleep_us(100);
        }

        uint16_t remaining = len - sent;
        uint16_t chunk_size = (remaining > free_size) ? free_size : remaining;

        uint8_t ptr_buf[2] = {0};
        read_sn(sn, 0x0024, ptr_buf, 2); // Read current Sn_TX_WR
        uint16_t wr_ptr = (uint16_t)((ptr_buf[0] << 8) | ptr_buf[1]);
        uint16_t offset = wr_ptr & buf_mask;

        if (offset + chunk_size <= buf_size) {
            cs_select();
            uint8_t hdr[3] = { (uint8_t)(offset >> 8), (uint8_t)(offset & 0xFF), tx_w_cb };
            spi_write_blocking(SPI_PORT, hdr, 3);
            spi_write_blocking(SPI_PORT, data + sent, chunk_size);
            cs_deselect();
        } else {
            uint16_t size1 = buf_size - offset;
            uint16_t size2 = chunk_size - size1;

            cs_select();
            uint8_t hdr1[3] = { (uint8_t)(offset >> 8), (uint8_t)(offset & 0xFF), tx_w_cb };
            spi_write_blocking(SPI_PORT, hdr1, 3);
            spi_write_blocking(SPI_PORT, data + sent, size1);
            cs_deselect();

            cs_select();
            uint8_t hdr2[3] = { 0x00, 0x00, tx_w_cb };
            spi_write_blocking(SPI_PORT, hdr2, 3);
            spi_write_blocking(SPI_PORT, data + sent + size1, size2);
            cs_deselect();
        }

        uint16_t new_wr = wr_ptr + chunk_size;
        uint8_t new_wr_buf[2] = { (uint8_t)(new_wr >> 8), (uint8_t)(new_wr & 0xFF) };
        write_sn(sn, 0x0024, new_wr_buf, 2);

        cmd_sn(sn, CR_SEND);

        start_t = time_us_64();
        while (1) {
            uint8_t cmd_reg = 0;
            read_sn(sn, 0x0001, &cmd_reg, 1);
            if (cmd_reg == 0) break;
            if (time_us_64() - start_t > 100000ULL) break;
            sleep_us(50);
        }

        sent += chunk_size;
    }
}

void w5500_disconnect_socket(uint8_t sn) {
    // 1. Check Sn_TX_RD == Sn_TX_WR to verify all TX bytes sent over network (up to 1000ms wait)
    uint64_t start_t = time_us_64();
    while (time_us_64() - start_t < 1000000ULL) {
        if (is_tx_buffer_empty(sn)) break;
        sleep_us(100);
    }
    sleep_ms(10);

    // 2. Issue graceful DISCON (TCP FIN) so browser cleanly receives 200 OK
    cmd_sn(sn, CR_DISCON);
}

void w5500_open_udp_socket(uint8_t sn, uint16_t port) {
    uint8_t status = w5500_get_socket_status(sn);
    if (status != SOCK_UDP) {
        w5500_close_socket(sn);
        uint8_t mode = 0x02; // UDP
        write_sn(sn, 0x0000, &mode, 1);
        uint8_t port_buf[2] = { (uint8_t)(port >> 8), (uint8_t)(port & 0xFF) };
        write_sn(sn, 0x0004, port_buf, 2);

        cmd_sn(sn, CR_OPEN);
        uint64_t start_t = time_us_64();
        while (time_us_64() - start_t < 100000ULL) {
            if (w5500_get_socket_status(sn) == SOCK_UDP) break;
            sleep_us(50);
        }
    }
}

uint16_t w5500_recv_udp_packet(uint8_t sn, uint8_t remote_ip[4], uint16_t *remote_port, uint8_t *buf, uint16_t max_len) {
    uint16_t avail = w5500_rx_bytes_available(sn);
    if (avail < 8) return 0; // UDP Header = IP(4) + Port(2) + Length(2)

    uint8_t header[8] = {0};
    w5500_read_rx_data(sn, header, 8);

    memcpy(remote_ip, header, 4);
    *remote_port = (uint16_t)((header[4] << 8) | header[5]);
    uint16_t payload_len = (uint16_t)((header[6] << 8) | header[7]);

    uint16_t read_len = (payload_len > max_len) ? max_len : payload_len;
    if (read_len > 0) {
        w5500_read_rx_data(sn, buf, read_len);
    }
    return read_len;
}

void w5500_send_udp_packet(uint8_t sn, const uint8_t remote_ip[4], uint16_t remote_port, const uint8_t *data, uint16_t len) {
    if (remote_ip[0] == 255 && remote_ip[1] == 255 && remote_ip[2] == 255 && remote_ip[3] == 255) {
        static const uint8_t bcast_mac[6] = {0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF};
        write_sn(sn, 0x0006, bcast_mac, 6);
    }
    write_sn(sn, 0x000C, remote_ip, 4);
    uint8_t port_buf[2] = { (uint8_t)(remote_port >> 8), (uint8_t)(remote_port & 0xFF) };
    write_sn(sn, 0x0010, port_buf, 2);
    w5500_send_tx_data(sn, data, len);
}

#endif // (TARGET_ETH_CHIP == CHIP_W5500)
